stablehlo/dialect/TypeInference.cpp

```diff
@@ -866,15 +867,24 @@ LogicalResult inferPadOp(
       return emitOptionalError(
           location,
           "Interior padding cannot be negative: ", paddingInteriorVal);
-    int64_t expectedOutput =
-        inputShape[i] + paddingLowVal + paddingHighVal +
-        std::max<int64_t>(inputShape[i] - 1, 0LL) * paddingInteriorVal;
-    if (expectedOutput < 0)
-      return emitOptionalError(
-          location, "Padding result in negative size for dimension ", i);
-    resultShape.push_back(expectedOutput);
+
+    if (!hlo::isDynamicDimSize(inputShape[i])) {  // static dim
+      int64_t expectedOutput = inferPadDim(inputShape[i], paddingLowVal,
+                                           paddingHighVal, paddingInteriorVal);
+      if (expectedOutput < 0)
+        return emitOptionalError(
+            location, "Padding result in negative size for dimension ", i);
+      resultShape.push_back(expectedOutput);
+
+      if (!inputBounds.empty())
+        resultBounds.push_back(ShapedType::kDynamicSize);
+    } else  // dynamic dim with bounds
+      resultBounds.push_back(inferPadDim(inputBounds[i], paddingLowVal,
```

Comment:
add check for negative on bounds as well?

Reply:
Sure! I was thinking of this once, and glad you point it out! Added.